File: src/string-calculator.ts

const setDelimiters = (expression: string): [string[], string] => {
  let delimiters = [',', '\n']; // Default delimiters
  if (expression.startsWith('//')) {
    expression = expression.slice(2); // '//'
    const current_char = expression[0];
    if (current_char !== '[') { // Unique delimiter
      delimiters = [current_char];
      expression = expression.slice(1);
      if (expression.startsWith('\n')) {
        expression = expression.slice(1);
      } else {
        throw Error('Unique delimiter ' + delimiters[0]
            + ' must be separated from the evaluated expression by a newline');
      }
    } else { // Multiple delimiters
      delimiters = getMultipleDelimiters(expression, []);
      expression = String(expression.split('\n').slice(1));
    }
  }
  return [
    delimiters,
    expression
  ]
};

const getMultipleDelimiters = (expression: string,
    delimiters: Array<string>) => {
  expression = expression.slice(1); // '['
  const delimiter = getDelimiter(expression, '');
  expression = expression.slice(delimiter.length + 1); // delimiter ']'
  delimiters.push(delimiter);
  if (expression[0] === '[') {
    return getMultipleDelimiters(expression, delimiters);
  } else {
    return delimiters;
  }
};

const getDelimiter = (expression: string, acc: string) => {
  const current_char = expression[0];
  if (current_char !== ']') {
    return getDelimiter(expression.slice(1), acc + current_char);
  } else {
    return acc;
  }
};

const isNumericChar = (char: string) => char === '-' || (char >= '0' && char <= '9');

const skipDelimiter = (expression: string, acc: string) => {
  if (isNumericChar(expression[1])) {
    return [expression, acc + expression[0]];
  } else {
    return skipDelimiter(expression.slice(1), acc + expression[0]);
  }
};

const recursiveStringCalculator = (expression: string,
    acc: string, delimiters: Array<string>, negatives: Array<number>) => {
  const isEmptyExpression = expression.length === 0;
  let currentChar = expression[0];
  const isDelimiterCurrentChar = !isNumericChar(currentChar);

  if (isDelimiterCurrentChar && currentChar) {
    const [expression_skipped, delimiter] = skipDelimiter(expression, '');
    expression = expression_skipped;
    currentChar = expression_skipped[expression_skipped.length - 1];
    if (!delimiters.includes(delimiter)) {
      throw Error('Delimiter ' + delimiter + ' is not in setted delimiters ' + delimiters);
    }
  }

  let result = 0;
  if (isEmptyExpression || isDelimiterCurrentChar) {
    const number_acc = Number(acc);
    if (number_acc < 0) {
      negatives.push(number_acc);
    }
    if (isEmptyExpression && negatives.length > 0) {
      throw Error('Negatives not allowed: ' + negatives);
    }
    result += number_acc <= 1000 ? number_acc : 0;
  }
  if (!isEmptyExpression) {
    let currentAcc = '';
    if (!isDelimiterCurrentChar) {
      currentAcc = acc + currentChar;
    }
    result += recursiveStringCalculator(expression.slice(1), currentAcc, delimiters, negatives);
  }
  return result;
};

export const stringCalculator = (expression: string) => {
  const [parsedDelimiters, parsedExpression] = setDelimiters(expression);
  return recursiveStringCalculator(parsedExpression, '', parsedDelimiters, []);
};

File: src/string_calculator.cpp

#include "../include/string_calculator.h"

StringCalculator::StringCalculator(void) {
  delimiters_ = {",", "\n"};
}

StringCalculator::~StringCalculator() {}

std::string StringCalculator::setDelimiters(std::string expression) {
  if (expression[0] == expression[1] && expression[1] == '/') {
    expression.erase(0, 2); // '//'
    if (expression[0] == '[') {
      delimiters_ = {};
      do {
        expression.erase(0, 1); // '['
        std::string delimiter = "";
        while (expression[0] != ']') {
          delimiter += expression[0];
          expression.erase(0, 1);
        }
        expression.erase(0, 1); // ']'
        delimiters_.push_back(delimiter);
      } while (expression[0] == '[');
    } else {
      delimiters_ = {std::string(1, expression[0])};
      expression.erase(0, 1); // delimiter
    }
    expression.erase(0, 1); // '\n'
  }
  return expression;
}

int StringCalculator::add(std::string expression) {
  expression = setDelimiters(expression);
  std::vector<std::string> splittedExpression = splitByDelimiters(expression);

  int result = 0;
  std::vector<int> negatives = {};
  for (int i = 0; i < splittedExpression.size(); i++) {
    int integer = std::stoi(splittedExpression[i]);
    if (integer < 0) {
      negatives.push_back(integer);
    }
    if (integer <= 1000) {
      result += integer;
    }
  }

  if (negatives.size() > 0) {
    std::string error = "Negatives are not allowed: ";
    for (int i = 0; i < negatives.size(); i++) {
      error += std::to_string(negatives[i]);
      if (i < negatives.size() - 1) {
        error += ", ";
      }
    }
    throw error;
  }

  return result;
}

std::vector<std::string> StringCalculator::splitByDelimiters(std::string expression) {
  std::string accumulator = "";
  std::vector<std::string> splitted = {};
  size_t i = 0;
  while (i < expression.size()) {
    bool foundDelimiter = false;
    for (size_t j = 0; j < delimiters_.size(); j++) {
      std::string currentDelimiter = "";
      std::vector<std::string> delimiters = delimiters_;
      for (size_t k = 0; k < delimiters_[j].size(); k++) {
        if (expression[i + k] != delimiters_[j][k]) {
          break;
        }
        currentDelimiter = delimiters_[j];
      }
      if (currentDelimiter != "") {
        foundDelimiter = true;
        splitted.push_back(accumulator);
        i += currentDelimiter.size();
        accumulator = "";
        break;
      }
    }
    if (!foundDelimiter) {
      accumulator += expression[i];
      i++;
    }
  }
  if (accumulator != "") {
    splitted.push_back(accumulator);
  }
  return splitted;
}
